Framework/Algorithms/src/MultipleScatteringCorrection.cpp

```diff
@@ -0,0 +1,403 @@
+                  "The number of wavelength points for which the numerical integral is\n"
+                  "calculated (default: all points)");
+
+  auto moreThanZero = std::make_shared<BoundedValidator<double>>();
+  moreThanZero->setLower(0.001);
+  declareProperty("ElementSize", 1.0, moreThanZero, "The size of one side of an integration element cube in mm");
+
+  declareProperty(std::make_unique<WorkspaceProperty<WorkspaceGroup>>("OutputWorkspace", "", Direction::Output),
+                  "Output workspace name. "
+                  "A Workspace2D containing the correction matrix that can be directly applied to the corresponding "
+                  "Event workspace for multiple scattering correction.");
+}
+
+/**
+ * @brief validate the inputs
+ *
+ * @return std::map<std::string, std::string>
+ */
+std::map<std::string, std::string> MultipleScatteringCorrection::validateInputs() {
+  std::map<std::string, std::string> result;
+
+  // check 0: input workspace must have a valida sample
+  // NOTE: technically the workspace validator should be able to catch the undefined sample
+  //       error.  Keeping this check here in case the validator is changed in the future.
+  m_inputWS = getProperty("InputWorkspace");
+  const auto &sample = m_inputWS->sample();
+  if (!sample.getShape().hasValidShape()) {
+    result["InputWorkspace"] = "The input workspace must have a valid sample shape";
+  }
+
+  // check 1: input workspace must have a valid sample environment
+  // TODO: this check should be implemented once we start considering the multiple
+  // scattering correction between container and sample (heterogenous scattering media)
+  // NOTE: use PaalmanPingAbsorptionCorrection.cpp check as starting point
+
+  // others?
+
+  return result;
+}
+
+/**
+ * @brief execute the algorithm
+ *
+ */
+void MultipleScatteringCorrection::exec() {
+  // parse input properties and assign corresponding values to the member
+  // variables
+  parseInputs();
+
+  // prepare the cached distances
+  MultipleScatteringCorrectionDistGraber distGraber(m_inputWS->sample().getShape(), m_elementSize);
+  distGraber.cacheLS1(m_beamDirection);
+  // NOTE: the following data is now cached in the distGraber object
+  // std::vector<double> distGraber.m_LS1 : Cached L1 distances
+  // std::vector<double> distGraber.m_elementVolumes : Cached element volumes
+  // std::vector<Kernel::V3D> distGraber.m_elementPositions : Cached element positions
+  // size_t distGraber.m_numVolumeElements : The number of volume elements
+  size_t numVolumeElements = distGraber.m_numVolumeElements;
+
+  // L2D needs to be calculated w.r.t the detector
+  // L12 is independent from the detector, therefore can be cached outside
+  // - L12 is a upper off-diagonal matrix
+  // NOTE: if the sample size/volume is too large, we might need to use openMP
+  //       to parallelize the calculation
+  size_t len_l12 = numVolumeElements * (numVolumeElements - 1) / 2;
```

Comment:
```suggestion
  const size_t len_l12 = numVolumeElements * (numVolumeElements - 1) / 2;
```
and the compiler will essentially get rid of this variable.

Reply:
good point, adding it now